Added more in-depth check for zone determination when dragging around

ebbyd3 opened the pull request:
the screen. Previously, it would iterate through the zones in the order
they were added and find the first one that fit the description. While
this works in most cases, if a user wants to have overlapping zones, it
is better to iterate through all of them and find the zone that the user
expects. There are cases where a zone is completely inaccessible on drag
because of the current code. To resolve this, the zone search will look
for the smallest zone possible. The reason I chose this solution is
because this guarantees that zones are at least reachable since if a
zone was bigger than another zone, then there must be a part of it
that is exposed, therefore reachable itself. Note: this solution is for
the scenario between two zones. More than that is not guaranteed. But I
feel like this covers enough scenarios to warrant its addition.

Example:
  ----------------
  - Zone1        -
  -  ----------  -
  -  - Zone2  -  -
  -  -        -  -
  -  ----------  -
  ----------------

Previously, zone2 was inaccessible since it would iterate through 1 then
2. But 1 would always be seen first when dragging a window. With this
fix it zone2 will be accessible.

<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [ ] Closes #xxx
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

ebbyd3 committed:
Added more in-depth check for zone determination when dragging around
the screen. Previously, it would iterate through the zones in the order
they were added and find the first one that fit the description. While
this works in most cases, if a user wants to have overlapping zones, it
is better to iterate through all of them and find the zone that the user
expects. There are cases where a zone is completely inaccessible on drag
because of the current code. To resolve this, the zone search will look
for the smallest zone possible. The reason I chose this solution is
because this guarantees that zones are at least reachable since if a
zone was bigger than another zone, then there must be a part of it
that is exposed, therefore reachable itself. Note: this solution is for
the scenario between two zones. More than that is not guaranteed. But I
feel like this covers enough scenarios to warrant its addition.

Example:
  ----------------
  - Zone1        -
  -  ----------  -
  -  - Zone2  -  -
  -  -        -  -
  -  ----------  -
  ----------------

Previously, zone2 was inaccessible since it would iterate through 1 then
2. But 1 would always be seen first when dragging a window. With this
fix it zone2 will be accessible.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/PowerToys?pullRequest=438) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/PowerToys?pullRequest=438) before we can accept your contribution.<br/><br/>:x: ebbyd3 [sign now](https://cla.opensource.microsoft.com/microsoft/PowerToys?pullRequest=438)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/PowerToys?pullRequest=438) it.</sub>

enricogior commented:
Hi @ebbyd3 
thanks for the PR!

enricogior commented:
Hi @ebbyd3
thank you for your patience, I finally had the time to review the PR.
It works perfectly, definitely a nice improvement!
I was wondering if calculating the `knownZoneArea` when the `smallestKnownZone` is set would be better than calculating it for each iteration. But that is just a nit.
I'm going to add a couple of comment on the code formatting, other than that it LGTM!

enricogior commented on src/modules/fancyzones/lib/ZoneSet.cpp:
```diff
@@ -77,17 +77,33 @@ IFACEMETHODIMP ZoneSet::RemoveZone(winrt::com_ptr<IZone> zone) noexcept
+            if (PtInRect(newZoneRect, pt))
             {
-                return zone;
+                if(smallestKnownZone == nullptr)
```
As a rule of thumb, let's use the original file formatting style. So please add a space after `if`. Thanks.

enricogior commented on src/modules/fancyzones/lib/ZoneSet.cpp:
```diff
@@ -77,17 +77,33 @@ IFACEMETHODIMP ZoneSet::RemoveZone(winrt::com_ptr<IZone> zone) noexcept
+                    int newZoneArea = (newZoneRect->right-newZoneRect->left)*(newZoneRect->bottom-newZoneRect->top);
+
+                    if(newZoneArea<knownZoneArea)
+                        smallestKnownZone = zone;
```
Also please put `smallestKnownZone = zone;` in curly braces. Thanks.

ebbyd3 committed:
Fixed some minor code styling issues. Removed some minor redundant calculations for a small performance enhancement with zone detection.

ebbyd3 commented:
Thanks for you feedback, @enricogior 

I've added in your suggestions for styling and also removed the knownArea calculation from the loop as you suggested. I didn't replace it though because I still have to return a pointer. Other than that looks like we are good to go.

enricogior commented:
Thanks @ebbyd3!
Contributions like yours are vital to make PowerToys awesome!
